### DISCUSSION TITLE
Glitch occurs when trying to view glTF files on certain GPUs

### DISCUSSION THREAD

**Discussion by cx20** [google/filament#3670] [Q&A]
I tried to display the following sample in a different GPU environment.
https://cx20.github.io/webgl-test/examples/filament/complex/index.html

Oddly enough, some environments can display correctly and some do not.
Is this a GPU issue? Or is it a library issue?

[OK] [Windows 10 + Chrome 89 + Filament v1.9.16 + GeForce GTX 660 result:](https://cx20.github.io/webgl-test/examples/filament/complex/index.html)
![image](https://user-images.githubusercontent.com/1833979/111820249-9d8ca400-8924-11eb-8fcd-d7f739d03e50.png)

[NG] [Windows 10 + Chrome 89 + Filament v1.9.16 + GeForce RTX 2060 result:](https://cx20.github.io/webgl-test/examples/filament/complex/index.html)
![image](https://user-images.githubusercontent.com/1833979/111820128-7930c780-8924-11eb-8de0-bbc1bafe7838.png)

**Comment by romainguy** [comment]
I can try on my RTX 3090 later, but it's probably a driver bug we're running into.

**Reply by cx20** [reply]
Thank you for your quick answer.

I don't know if it's related, I checked the console again and noticed the following warning message (I don't think there was a warning in the console when I checked it about 30 minutes ago)

```
index.html:1 
[.WebGL-000050000013BB80]GL Driver Message (OpenGL, Performance, GL_CLOSE_PATH_NV, Medium): You could get a better shader compiling performance if you re-write the uniform block(s)
[ FrameUniforms ]
in the FRAGMENT shader.
You could get more details from https://chromium.googlesource.com/angle/angle/+/refs/heads/master/src/libANGLE/renderer/d3d/d3d11/UniformBlockToStructuredBufferTranslation.md
```

**Reply by romainguy** [reply]
Not sure what UBO it would be since it apparently happens with arrays >= 50. @pixelflinger any idea? I doubt we would bother rewriting our UBOs to make the GL->DirectX codepath happy though :p

**Reply by romainguy** [reply]
Windows 10 + Chrome 88 + RTX 3090, a similar bug happens at first but as soon as the T-Rex model finishes loading, everything looks fine.

**Reply by cx20** [reply]
In my Chrome 10 + RTX 2060 environment, Glitch doesn't seem to improve even if T-Rex has finished loading.

BTW, it may not be relevant, but when I display this sample page in Firefox 86, it seems that the following error is displayed on the console.

```
filament.wasm:413982:1
Uncaught RuntimeError: index out of bounds
    dynCall_viiif https://cx20.github.io/gltf-test/libs/filament/v1.9.16/filament.js:9
    dynCall_viiif_774 https://cx20.github.io/gltf-test/libs/filament/v1.9.16/filament.js line 9 > Function:4
    gltfio$Animator$applyAnimation https://cx20.github.io/gltf-test/libs/filament/v1.9.16/filament.js line 9 > Function:10
    render https://cx20.github.io/webgl-test/examples/filament/complex/index.js:140
```
​

**Reply by cx20** [reply]
I found an error in my sample program. Setting the Clear Options in the renderer solved the glitching problem.
https://github.com/cx20/webgl-test/commit/b2397efae4d4f5f08abd60f2fc00c9864052c023
![image](https://user-images.githubusercontent.com/1833979/111872950-5c060280-89d1-11eb-9916-5e7e81768c3e.png)

**Reply by cx20** [reply]
I have tried some official Filament samples on Chrome 89 + `RTX 2060`.
Perhaps the reason the Triangle tutorial doesn't glitch is because `setClearOptions` is called.
I think the reason why the knotess sample does not have any problems is probably because the version of Filament is old.

|OK/NG|Sample name|result|
|:----:|:----|:----:|
|OK|tutorials/[Triangle Tutorial](https://google.github.io/filament/webgl/tutorial_triangle.html)|![image](https://user-images.githubusercontent.com/1833979/111875842-28ca7000-89df-11eb-8a86-0e2e52f1f0cf.png)|
|NG|tutorials/[Redball Tutorial](https://google.github.io/filament/webgl/tutorial_redball.html)|![image](https://user-images.githubusercontent.com/1833979/111875856-45ff3e80-89df-11eb-9a11-b986bbca84a4.png)|
|NG|tutorials/[Suzanne Tutorial](https://google.github.io/filament/webgl/tutorial_suzanne.html)|![image](https://user-images.githubusercontent.com/1833979/111875872-56afb480-89df-11eb-820b-0cfba61410a3.png)|
|NG|demos/[parquet](https://google.github.io/filament/webgl/parquet.html)|![image](https://user-images.githubusercontent.com/1833979/111875885-6b8c4800-89df-11eb-94fd-828decbc3cdb.png)|
|NG|demos/[suzanne](https://google.github.io/filament/webgl/suzanne.html)|![image](https://user-images.githubusercontent.com/1833979/111875894-7515b000-89df-11eb-9778-08fc26a36e89.png)|
|NG|demos/[helmet](https://google.github.io/filament/webgl/helmet.html)|![image](https://user-images.githubusercontent.com/1833979/111875908-7e9f1800-89df-11eb-9939-7cf6e0dd9f8c.png)|
|OK|demos/[knotess](https://prideout.net/knotess/)|![image](https://user-images.githubusercontent.com/1833979/111875920-8b237080-89df-11eb-94c7-7d2d2a1c22a2.png)|

**Reply by cx20** [reply]
Probably the same problem as https://github.com/google/filament/issues/3151.

**Comment by romainguy** [comment]
@pixelflinger @prideout These demos don't need to clear the color buffer but I wonder if we have a bug with clearing the depth buffer? It's likely related to framegraph2.
